### PR TITLE
FTEQW Config Overhaul

### PR DESCRIPTION
**Pull request by Zungrysoft**
This is an overhaul of the default configuration in order to support FTEQW as our primary recommended game engine. I think FTEQW is ideal because it supports both singleplayer and deathmatch very well and has a ton of customization options we can use to fine-tune the default LibreQuake experience. We should make sure LibreQuake is compatible with most modern engines, especially the popular ones. But our focus for compatibility should be FTEQW.

Ironwail is still a good secondary choice so I made sure this config works well for it too. But Ironwail does not have good deathmatch support which is why FTEQW should be the first choice.

All changes:
- Customized FTEQW configuration to roughly match the look LibreQuake has with Ironwail.
- Added graphical style toggle.
	- Quake essentially has two artstyles at this point: The modern QuakeSpasm style most community maps have been made for and the retro DOS Quake style the game had when it released. I think it's worth celebrating Quake's history by having both of these graphics styles available at the press of a button. Pressing F10 will toggle between two styles:
		- Modern: The look of really new engines like FTEQCC and the Quake Rerelease.
		- Retro: Software rendered look similar to DOS Quake or WinQuake.
	- These styles are defined in unpacked cfg files, which makes it easy for users to further customize these styles to their personal taste.
- Bound Mouse2 to Jump. This is a common bind for deathmatch players.
- Changed accessibility.md to accessibility.txt to make it more accessible. Added some new information.
- Disabled automatic load.
	- LibreQuake singleplayer levels are designed similarly to the levels in the base game, meaning they should all be reasonable to complete without using the Save/Load feature. 
	- Automatic saving is still enabled. Though player will have to manually load the save.
	- Automatic saving fundamentally changes the way the game plays. It should be opt-in, not opt-out.
- Cleaned up formatting in default.cfg
- Added .vscode folder to gitignore.
- Changed taskbar string from 'Quake' to 'LibreQuake'
- Updated README with new installation instructions.

I am open to any feedback on these changes, since some of them are rather subjective.

Big thanks to Spike for helping me with all of this config work!

Modern style:
fte-20240915015913-0

Retro Style:
fte-20240915015917-0

### CODE REVIEW

File: lq1/accessibility.txt
Comment (MotoLegacy) [inline lq1/accessibility.txt:5]: Do you mean to say FTEQW, not FTEQCC?

File: lq1/accessibility.txt
Comment (MotoLegacy) [inline lq1/accessibility.txt:11]: Which engines support using TrueType fonts like this?

File: lq1/default.cfg
Comment (MotoLegacy) [inline lq1/default.cfg:130]: Shouldn't this be in an ironwail specific config file?

File: lq1/default.cfg
Comment (MotoLegacy) [inline lq1/default.cfg:142]: Ditto, does ezquake have its own cfg it executes?

File: lq1/style_modern.cfg
Comment (MotoLegacy) [inline lq1/style_modern.cfg:1]: FTEQW?

File: lq1/style_retro.cfg
Comment (MotoLegacy) [inline lq1/style_retro.cfg:1]: FTEQW?

**Review by MotoLegacy — CHANGES_REQUESTED** [review]
I'm not sure how I feel about jamming different cvars for different engines into single configs. I think there's a better way we can approach this using the fact most engines load configs specified by their engine name, using them to create general aliases. I suspect that would be a lot more elegant and scaleable in case this is going to be done again.

File: lq1/accessibility.txt
Comment (Zungrysoft) [reply lq1/accessibility.txt]: Dammit! I keep doing that! I removed that line anyway since I was able to make the command FTEQW compatible.

File: lq1/accessibility.txt
Comment (Zungrysoft) [reply lq1/accessibility.txt:11]: I don't have an exhaustive list. I know FTEQW supports it and Ironwail doesn't. Those are the only two engines I've tested so far. Spoike commented that LQ's font is not very readable and suggested this command. I figured it would be a good entry in accessibility.

File: lq1/default.cfg
Comment (Zungrysoft) [reply lq1/default.cfg:131]: Does Ironwail have its own config file other than ironwail.cfg? ironwail.cfg isn't meant to be distributed since it gets automatically written by the engine to save your settings. Also, a lot of these commands work in other QuakeSpasm forks.

File: lq1/default.cfg
Comment (Zungrysoft) [reply lq1/default.cfg:143]: See my above comment.

**Comment by Zungrysoft** [comment]
@MotoLegacy This is the way we've been doing it before. The alternative is putting the settings in config.cfg/ironwail.cfg/fte.cfg/etc. But those files get overwritten by the engine every time you load up so I don't think it's safe to put this sort of default configuration in those files. This config should be in default.cfg. If these engines have their own cfg file that doesn't get written by the engine, that would work.

File: lq1/default.cfg
Comment (MotoLegacy) [reply lq1/default.cfg:131]: `ironwail.cfg` is just a sub for `config.cfg`. Yes, it is overwritten when you apply new settings, it is my understanding though that it should preserve aliases?

It's definitely worth investigating organizing these better. I know it's not a problem you've created but it is a mess. It may not "break" anything but we can't actually be sure and we know it will cause really issues on many engines where it's forced to go through all of these unregistered cvars and spamming the console. 

If we had a better backbone for this it would make supporting enhanced features of other engines significantly more streamlined both if we ever want to create more configs/aliases ourselves and if outside contributors want to add that support.

**Comment by Nolcoz** [comment]
Okay, so I tested it on the vulkan version of fteqw and I have a lot of gripes with this. Mostly the engine that is.
As for the FTEQW config this is what I think:
- Hud scale 2x is better I think. Hud scale 3x is just way too big to be honest. To the point that some text clips out of the screen and you have to scroll down in some menus.
- Making the hud slightly transparent? If that is possible that is, would help a little with visibility.
- Is there no way to make it so animation interpolation isnt applied to guns too? On the lq weapons its mostly fine (the super nailgun kinda looks worse with the interpolated spinning in my opinion) but with weapons from mods it can very easily look weird, taking the id1 nailgun as an example.
 >Automatic saving is still enabled. Though player will have to manually load the save.
- How does this work? Like, is it a bind that has to be made

Outside of that I think the configs are fine.
Okay, now the engine. I am gonna be honest, I dont like. For multiplayer I can see it being good, but after testing it with some maps I have so many problems with it.
- Fog seems to work completly different than in engines derived from Quakespasm, where it seems to be way closer to the player but at the same time it isnt as heavy. After testing why this happens it seems to be related to the gamma and contrast settings I have. Even then I still havent been able to repliacte the quakespasm or ironwail fog without the menu looking way to contrasted or the game looking too flat.
These pics were taken with the default values in the cfg.
Here are some examples with e1m7(black fog)
Ironwail
e1m7_2024-09-15_14-48-17
FTEQW
fte-20240915140932-0
e3m3(white fog)
Ironwail
e3m3_2024-09-15_14-52-37
FTEQW
fte-20240915142927-0

- Lit water doesnt work. I know its minor, but this feels like such a huge stepback for the look of some levels, and with most modern quake maps supporting this feature, it really feels like a downgrade in that regard.
Ironwail
e3m3_2024-09-15_14-53-30
FTEQW
fte-20240915143022-0

- Weird performance stuff? I am using the Vulkan version of fteqw so I dont know if that influenced on performance but(this also happened when not recording for reference).
https://github.com/user-attachments/assets/4f5db426-cd75-43bc-9631-673a0edfa087

- Some maps are just broken.
In the end of e1m7, these enemies are supposed to be on a clip platform and each enemy is placed on a trigger hurt that will kill them instantly, that later descends so the bodies are on the floor. In FTEQW these enemies are alive
FTEQW
fte-20240915142331-0
Quakepasm
spasm0002
Also this in FTEQW, which is very much not meant to happen in Quakespasm nor Ironwail
https://github.com/user-attachments/assets/548d6b6f-210a-469e-a824-4d3c8b33a19a

- More of a personal thing but I also had some weird issues with Vsync where it would make mouse movement really slugish.

Ye, thats why I think about it. I do know that for multiplayer this is THE engine, but for singleplayer I have a lot of reservations with it.

**Comment by MotoLegacy** [comment]
Making a note here that I have had an off-platform discussion with Zungry and Spike and we have come up with a solution for engine identification to clean up the configuration files.

File: lq1/default.cfg
Comment (lavenderdotpet) [reply lq1/default.cfg:131]: > Shouldn't this be in an ironwail specific config file?

we don't ship engine specific configs so no

**Comment by Zungrysoft** [comment]
@Nolcoz Thanks so much for your feedback! Let me answer your points one by one:

- Personally, I think 3 fits the style better. 2 just feels too small. The user can easily change this using `scr_conscale`. Maybe the correct solution to this is to use `cl_sbar 0` for the modern style, so the size of the HUD doesn't get in the way.
- For animation interpolation, I don't know if there is anything similar to Quakespasm's "pause animation for 2 frames after a muzzle flash" feature. It has r_lerpmuzzlehack, but it works a bit different and actually makes LibreQuake's weapon animations look way worse.
- For automatic saving, when I was testing it would automatically create save files every so often. I haven't seen it do it recently though. Not sure what changed. So saving is manual, just like in the base game.
- As far as graphical inconsistencies go, I compared my install of fteqw to your screenshots and my game looks much closer to your Ironwail screenshots than your FTEQW screenshots. Lit water works perfectly fine for me. Maybe it's a Vulkan thing? I used the FTEQW download from here: https://fte.triptohell.info/downloads
- For the broken maps, I think it would be best to fix the maps. I noticed the e2m3 bug as well in my testing. I'm sure with some fiddling, we can get everything to work in both engines. I'll look into the specific issues you identified and see if there are easy fixes for them.

**Comment by lavenderdotpet** [comment]
once this is done please wait for me to review it

File: lq1/style_modern.cfg
Comment (MotoLegacy) [inline lq1/style_modern.cfg:1]: Can you make these cvars just an alias under our new engine-specific configs? This will keep things more engine-organized and also won't have the side effect of making people think it's broken if they're using some other engine and expecting the behavior to work.

File: lq1/style_retro.cfg
Comment (MotoLegacy) [inline lq1/style_retro.cfg:1]: Ditto

**Comment by MotoLegacy** [comment]
Much nicer, thanks for addressing my points. Left another small comment and then I'll happily approve.

File: lq1/style_modern.cfg
Comment (Zungrysoft) [reply lq1/style_modern.cfg:1]: That's how I had it at first. I put them in separate config files so that if the user decides that they want to change the way modern style or classic style works, they can just edit those files. Like, maybe they want the classic HUD on modern style because they like the way the classic HUD looks. They can just go into style_modern.cfg and change `cl_sbar` from 0 to 1. With aliases, they would have to completely rewrite those aliases from scratch.

One alternative I considered is having another alias called `lq_style_modern_user` which the user can write themselves and include in autoexec.cfg. And that always gets called after `lq_style_modern_exec`. Then we provide a template autoexec.cfg which includes those alias commands, starting empty.

Comment (MotoLegacy) [reply lq1/style_modern.cfg:1]: `default_fte.cfg: alias style_modern="exec lq_style_modern_fte.cfg"`
etc. is what I had in mind. If that helps clarify. We just want the multi-engine stuff consolidated down.

**Comment by Zungrysoft** [comment]
@lavenderdotpet Spike politely asks why vid_restart is in default.cfg. He says it is unnecessary since the engine should already call vid_restart anyway.

https://github.com/lavenderdotpet/LibreQuake/commit/4e8a8e6ee

**Comment by Nolcoz** [comment]
>As far as graphical inconsistencies go, I compared my install of fteqw to your screenshots and my game looks much closer to your Ironwail screenshots than your FTEQW screenshots. Lit water works perfectly fine for me. Maybe it's a Vulkan thing? I used the FTEQW download from here: https://fte.triptohell.info/downloads

After downloading it from there, pretty much all graphical problems I had where gone and just looks the same as my Ironwail setup. You really did a perfect job at replicating that look.

>For the HUD scale, I personally, I think 3 fits the style better. 2 just feels too small. The user can easily change this using scr_conscale, which will get saved into their engine config file. I changed the HUD style for the "modern" style to use the more modern HUD design in FTEQW and Ironwail so the size of the HUD doesn't get in the way.

Checking your new version, it is much better ye. I think something like 2.5 would be a nice middle ground but it is a very minor thing tho now.

>For animation interpolation, I don't know if there is anything similar to Quakespasm's "pause animation for 2 frames after a muzzle flash" feature. It has r_lerpmuzzlehack, but it works a bit different and actually makes LibreQuake's weapon animations look way worse. It's tough because there probably are modded weapons that were made for Quakespasm's lerping rules. But I think this is the best way to go.

Ahh I see. Well I thats fine then.

>For the broken maps, I think it would be best to fix the maps. I noticed the e2m3 bug as well in my testing. I'm sure with some fiddling, we can get everything to work in both engines. I'll look into the specific issues you identified and see if there are easy fixes for them.

K then. I can imagine the e1m7 problem can be fixed by making it so the brush that lowers the corpses isnt a clip brush.
As for the key in e2m3, not sure what the problem is, but a similar technicque is used in e3m4 for armor and that seems to work fine, so it might be possible to fix

>For automatic saving, the engine will automatically save the game every few minutes or so. It will notify you in console whenever it autosaves. You can manually load it through the Save/Load menu as normal.

Tried it and it seems to work like you said ye.

With all that in mind, things seem to work pretty well now. Only real gripe now is the scale of the hud but again its a minor thing. If you want to keep that scale just tell me and I will aprove this, all seems well.
Also sorry if I was condescensding to FTE before, was just tired thats all

**Comment by Zungrysoft** [comment]
@Nolcoz I think I'm going to keep the HUD size the same. I still think 2 is too small, and I don't think decimal sizes look great either.

I'll look into all these map issues and try to identify fixes for all of them. There are some issues in my own maps as well.

I was pretty frustrated with FTE myself when I first started. Then I went into the FTE Discord and Spike helped me fix all of my problems. Turns out this engine is pretty amazing when you actually know what the commands for things are lol.

**Review by MotoLegacy — APPROVED** [review]
Looks great!

**Comment by Nolcoz** [comment]
>I was pretty frustrated with FTE myself when I first started. Then I went into the FTE Discord and Spike helped me fix all of my problems. Turns out this engine is pretty amazing when you actually know what the commands for things are lol.
Ah okay, thats nice to hear. Ye, the first time I used FTE wasnt really pleasent, but with your configs I do like it way more now.

>I'll look into all these map issues and try to identify fixes for all of them. There are some issues in my own maps as well.
Should I wait for that before approving, or is that more of a future problem more than anything?

**Comment by MotoLegacy** [comment]
@Nolcoz 
You should open issues concerning them so we can track it and fix it when you can. I don't think they should be a blocker for this PR.

**Review by Nolcoz — APPROVED** [review]
Ok then. Its all good here

**Comment by lavenderdotpet** [comment]
> @lavenderdotpet Spike politely asks why vid_restart is in default.cfg. He says it is unnecessary since the engine should already call vid_restart anyway.
> 
> You made the change in this commit: https://github.com/lavenderdotpet/LibreQuake/commit/4e8a8e6ee

I don't remember feel free to remove

File: lq1/qwprogs.dat
Comment (MotoLegacy) [inline lq1/qwprogs.dat:1]: I assume this was a mistake, can you remove this?

**Review by MotoLegacy — CHANGES_REQUESTED** [review]
See comment

File: lq1/qwprogs.dat
Comment (Zungrysoft) [reply lq1/qwprogs.dat:1]: Whoops. Thanks for the catch.

**Review by MotoLegacy — APPROVED** [review]

**Comment by lavenderdotpet** [comment]
image

retro should prob have no crosshair

**Comment by Zungrysoft** [comment]
@lavenderdotpet

I considered turning off the crosshair for the retro style. I decided against it because first, I think it's more of a gameplay feature than a graphics feature and second, crosshair was an option even in the original DOS Quake. It was just off by default. I think disabling the crosshair for retro mode would be kind of like disabling mouselook.

As for the viewmodel_quake, I really don't think it looks better in either mode, especially with LQ's viewmodels. But that's just my personal opinion.

**Comment by lavenderdotpet** [comment]
> As for the viewmodel_quake, I really don't think it looks better in either mode, especially with LQ's viewmodels. But that's just my personal opinion.

i prefer how r_viewmodel_quake looks due to the nice movement u get when looking up and down
also why does ur retro style shotgun look so long 


https://github.com/user-attachments/assets/0ac99c5c-00c3-4fe1-8499-82770cb40048

**Comment by Nolcoz** [comment]
>i prefer how r_viewmodel_quake looks due to the nice movement u get when looking up and down

I second this ye. Makes it feel more dinamic and less like its an image stuck to the middle of the screen

**Comment by Zungrysoft** [comment]
@lavenderdotpet @Nolcoz Alright. Since I'm outvoted here, I'll add it in. My problem was that it makes the viewmodel too far forward on the screen, but I can compensate by using viewmodel_fov of 80.

**Comment by lavenderdotpet** [comment]
> @lavenderdotpet @Nolcoz Alright. Since I'm outvoted here, I'll add it in. My problem was that it makes the viewmodel too far forward on the screen, but I can compensate by using viewmodel_fov of 80. I never noticed that it moves when you look up. That's actually pretty neat.

at a later point we can look at maybe moving the position of the model a little bit 

gonna do my test and review of this now

**Review by lavenderdotpet — COMMENTED** [review]
jeez what a typo i made with calling fteqw fteqcc

**Comment by lavenderdotpet** [comment]
>Automatic saving is still enabled. Though player will have to manually load the save.
>Automatic saving fundamentally changes the way the game plays. It should be opt-in, not opt-out.
@Zungrysoft 

i think automatic load should be left on since people using engines that have it on by default will be confused when it suddenly stops working for them with librequake
either the engine dev is gonna get issues over this or we are 

i think if we wanna change that we should talk with the engine devs about it or leave it to default

**Comment by lavenderdotpet** [comment]
image

lq_toggle_style doesnt seem to be working its defaulting to the fte one and not the ironwail one

**Comment by lavenderdotpet** [comment]
https://github.com/user-attachments/assets/7bd06cdb-d796-4cce-9d9f-8631aa49de89

the uvs are changing between the 2 modes why is that?

**Comment by lavenderdotpet** [comment]
>- Changed taskbar string from 'Quake' to 'LibreQuake'

@Zungrysoft  could u elaborate on what this means

**Comment by Zungrysoft** [comment]
@lavenderdotpet If someone is used to the autosave feature, they'll probably be downloading the mod release which doesn't contain any of this config stuff anyway. For people downloading the full release, I strongly maintain that autosave should be off by default so we're not adding an unexpected feature that warps the way the game is played. Our levels are designed to be short enough to complete without save/load. And if someone wants save/load, they can still do it manually using F6 and F9. And the game still creates automatic saves. You just have to load them yourself. Just one extra step.

Compare this with the alternative. If this feature is not disabled, there is no way to restart the level from the beginning without losing all of your items. So people (like me) who don't play with save/load are completely screwed if they don't know FTEQW's hacky workaround to disable this feature.

I do think it would be nice if FTEQW had a cvar to control it like with Ironwail. Currently, it requires this hacky workaround that can't easily be disabled. I'll look at the codebase and see if I can open a PR for it. Because I do think it's good for people to have the option. But it definitely should not be the default.

**Comment by Nolcoz** [comment]
> the uvs are changing between the 2 modes why is that?
From previous testing, that seems to be a thing with the software renderer where the uvs of the gun close to the camera get distorted. Also happens with regular id1. Dont know much about the rendering side of quake so I could just be getting things confused tho

**Comment by lavenderdotpet** [comment]
> @lavenderdotpet If someone is used to the autosave feature, they'll probably be downloading the mod release which doesn't contain any of this config stuff anyway. For people downloading the full release, I strongly maintain that autosave should be off by default so we're not adding an unexpected feature that warps the way the game is played. Our levels are designed to be short enough to complete without save/load. And if someone wants save/load, they can still do it manually using F6 and F9. And the game still creates automatic saves. You just have to load them yourself. Just one extra step.
> 
> Compare this with the alternative. If this feature is not disabled, there is no way to restart the level from the beginning without losing all of your items. So people (like me) who don't play with save/load are completely screwed if they don't know FTEQW's hacky workaround to disable this feature.
> 
> I do think it would be nice if FTEQW had a cvar to control it like with Ironwail. Currently, it requires this hacky workaround that can't easily be disabled. I'll look at the codebase and see if I can open a PR for it. Because I do think it's good for people to have the option. But it definitely should not be the default.

agreed it shouldn't be on by default 

side question about this can cvars/aliases be executed by triggers 
like how u can change maps or skill with triggers 

im asking so if maybe we can have a settings section in our start map for this stuff

**Comment by lavenderdotpet** [comment]
> > the uvs are changing between the 2 modes why is that?
> 
> From previous testing, that seems to be a thing with the software renderer where the uvs of the gun close to the camera get distorted. Also happens with regular id1. Dont know much about the rendering side of quake so I could just be getting things confused tho

thing is tho with fte and ironwail it isnt actually using a software render only emulating parts of it

**Comment by Nolcoz** [comment]
> > > the uvs are changing between the 2 modes why is that?
> > 
> > 
> > From previous testing, that seems to be a thing with the software renderer where the uvs of the gun close to the camera get distorted. Also happens with regular id1. Dont know much about the rendering side of quake so I could just be getting things confused tho
> 
> thing is tho with fte and ironwail it isnt actually using a software render only emulating parts of it

I think its essentially emulating that effect then. It may be a thing with winquake too

**Comment by Zungrysoft** [comment]
@lavenderdotpet The UV's change because the Retro style emulates the software renderer's look, which uses affine texture projection for vertex models. If you've ever wondered why Quake uses BSP models for health and ammo packs instead of MDL models, this is why. If you use the software emulation mode in Copper (which uses MDL models for health and ammo instead of BSP models), this is what it looks like:
fte-20240927194030-0

It's super obvious on boxy models like health and ammo, but barely noticeable on most other models. Viewmodels are probably the most affected MDL models. The original game designed its viewmodels so you can't tell unless you're paying attention. It's most obvious on the Shotgun model.

It's a quirk to be sure, but these sorts of quirks are what give the software renderer its charm.

**Comment by lavenderdotpet** [comment]
@Nolcoz 
> I think its essentially emulating that effect then. It may be a thing with winquake too


https://github.com/user-attachments/assets/a01216f2-4755-45a9-aee0-c86ca0583883


yeah yall are right i miss remembered and thought they didnt emulate that part

**Comment by Zungrysoft** [comment]
@lavenderdotpet The taskbar string refers to what your operating system calls the program in your OS taskbar. So instead of it saying "Quake: e1m4" it will say "LibreQuake: e1m4"

image

**Comment by lavenderdotpet** [comment]
image
retro should have scr_sbaralpha set to 1 and not 0.9999

image
top  scr_sbaralpha0.9999
bottom scr_sbaralpha

**Comment by lavenderdotpet** [comment]
> @lavenderdotpet The taskbar string refers to what your operating system calls the program in your OS taskbar. So instead of it saying "Quake: e1m4" it will say "LibreQuake: e1m4"
> 
> image

oooo! heard i like that

**Comment by Zungrysoft** [comment]
@lavenderdotpet 

> side question about this can cvars/aliases be executed by triggers like how u can change maps or skill with triggers
> 
> im asking so if maybe we can have a settings section in our start map for this stuff

Not possible with the default progs.dat. We could have it as a feature in our custom progs.dat, but it would not work in any mods, so we would have to wall it off if the player is not using our progs.

**Comment by Zungrysoft** [comment]
> retro should have scr_sbaralpha set to 1 and not 0.9999

That's correct. I'll fix that.

**Comment by lavenderdotpet** [comment]
> @lavenderdotpet
> 
> > side question about this can cvars/aliases be executed by triggers like how u can change maps or skill with triggers
> > im asking so if maybe we can have a settings section in our start map for this stuff
> 
> Not possible with the default progs.dat. We could have it as a feature in our custom progs.dat, but it would not work in any mods, so we would have to wall it off if the player is not using our progs.

hmmm darn maybe just a room that tells u what all the aliases are that or we do the progs thing but if not using our progs the buttons for it dont show up so u would only see the aliases in text form

**Review by lavenderdotpet — APPROVED** [review]

**Comment by lavenderdotpet** [comment]
I think this should be good 
we should have make discussion about the settings room idea at some point but I think this is good

**Pull request merged**
